File: python/src/contam_studio_core/contam_semantic_draft.py
from __future__ import annotations

from dataclasses import dataclass
import hashlib
import json
import re
from typing import Any, Mapping


SCHEMA_VERSION = "contam_semantic_draft.v1"
MAX_ZONES = 256
MAX_FLOW_PATHS = 512

_ID = re.compile(r"^[A-Za-z0-9_.:-]{1,160}$")
_SHA256 = re.compile(r"^[A-Fa-f0-9]{64}$")
_CONTAM_NAME = re.compile(r"^[A-Za-z0-9_.-]{1,15}$")
_ROOT_KEYS = frozenset(
    {
        "schema_version",
        "status",
        "draft_id",
        "project_session_id",
        "identity_sha256",
        "source_sha256",
        "revision_id",
        "draft_revision",
        "provenance",
        "zones",
        "flow_paths",
    }
)
_ZONE_KEYS = frozenset(
    {
        "id",
        "level_number",
        "name",
        "display_name",
        "volume_litres",
        "volume_basis",
        "geometry_region_id",
        "initial_temperature_millikelvin",
        "initial_pressure_millipascal",
    }
)
_FLOW_PATH_KEYS = frozenset(
    {
        "id",
        "level_number",
        "opening_id",
        "from_endpoint",
        "to_endpoint",
        "flow_element_id",
        "multiplier_millionths",
        "x_mm",
        "y_mm",
        "relative_height_mm",
        "direction_degrees",
    }
)


class ContamSemanticDraftError(ValueError):
    def __init__(self, code: str) -> None:
        super().__init__(code)
        self.code = code


@dataclass(frozen=True, slots=True)
class DraftContext:
    project_session_id: str
    identity_sha256: str
    source_sha256: str
    revision_id: str
    level_numbers: frozenset[int]
    existing_zone_ids: frozenset[str]
    supported_flow_element_ids: frozenset[str]
    geometry: Mapping[str, Any] | None = None


@dataclass(frozen=True, slots=True)
class DraftValidation:
    status: str
    draft_sha256: str
    diagnostics: tuple[str, ...]


def _object(value: Any, expected: frozenset[str], code: str) -> Mapping[str, Any]:
    if not isinstance(value, dict) or frozenset(value) != expected:
        raise ContamSemanticDraftError(code)
    return value


def _string(value: Any, pattern: re.Pattern[str], code: str) -> str:
    if not isinstance(value, str) or pattern.fullmatch(value) is None:
        raise ContamSemanticDraftError(code)
    return value


def _integer(value: Any, minimum: int, maximum: int, code: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not minimum <= value <= maximum:
        raise ContamSemanticDraftError(code)
    return value


def _display_name(value: Any) -> str:
    if (
        not isinstance(value, str)
        or not 1 <= len(value) <= 80
        or value != value.strip()
        or any(ord(character) < 32 or ord(character) == 127 for character in value)
    ):
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    return value


def _endpoint(value: Any) -> Mapping[str, Any]:
    item = _object(value, frozenset({"kind", "zone_id"}), "semantic_draft_contract_invalid")
    if item["kind"] == "outdoor" and item["zone_id"] is None:
        return item
    if item["kind"] == "zone":
        _string(item["zone_id"], _ID, "semantic_draft_contract_invalid")
        return item
    raise ContamSemanticDraftError("semantic_draft_contract_invalid")


def _zone(value: Any) -> Mapping[str, Any]:
    item = _object(value, _ZONE_KEYS, "semantic_draft_contract_invalid")
    _string(item["id"], _ID, "semantic_draft_contract_invalid")
    _integer(item["level_number"], 1, 10_000, "semantic_draft_contract_invalid")
    _string(item["name"], _CONTAM_NAME, "semantic_draft_contract_invalid")
    _display_name(item["display_name"])
    _integer(item["volume_litres"], 1, 1_000_000_000, "semantic_draft_contract_invalid")
    if item["volume_basis"] not in {"explicit", "geometry_estimate_confirmed"}:
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    _string(item["geometry_region_id"], _ID, "semantic_draft_contract_invalid")
    _integer(
        item["initial_temperature_millikelvin"],
        173_150,
        373_150,
        "semantic_draft_contract_invalid",
    )
    _integer(
        item["initial_pressure_millipascal"],
        -1_000_000_000,
        1_000_000_000,
        "semantic_draft_contract_invalid",
    )
    return item


def _flow_path(value: Any) -> Mapping[str, Any]:
    item = _object(value, _FLOW_PATH_KEYS, "semantic_draft_contract_invalid")
    _string(item["id"], _ID, "semantic_draft_contract_invalid")
    _integer(item["level_number"], 1, 10_000, "semantic_draft_contract_invalid")
    _string(item["opening_id"], _ID, "semantic_draft_contract_invalid")
    _endpoint(item["from_endpoint"])
    _endpoint(item["to_endpoint"])
    _string(item["flow_element_id"], _ID, "semantic_draft_contract_invalid")
    _integer(
        item["multiplier_millionths"],
        1,
        1_000_000_000_000,
        "semantic_draft_contract_invalid",
    )
    _integer(item["x_mm"], -1_000_000_000, 1_000_000_000, "semantic_draft_contract_invalid")
    _integer(item["y_mm"], -1_000_000_000, 1_000_000_000, "semantic_draft_contract_invalid")
    _integer(item["relative_height_mm"], 0, 100_000_000, "semantic_draft_contract_invalid")
    _integer(item["direction_degrees"], -1, 359, "semantic_draft_contract_invalid")
    return item


def parse_contam_semantic_draft(value: Any) -> Mapping[str, Any]:
    root = _object(value, _ROOT_KEYS, "semantic_draft_contract_invalid")
    if root["schema_version"] != SCHEMA_VERSION or root["status"] != "available":
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    _string(root["draft_id"], _ID, "semantic_draft_contract_invalid")
    _string(root["project_session_id"], _ID, "semantic_draft_contract_invalid")
    _string(root["identity_sha256"], _SHA256, "semantic_draft_contract_invalid")
    _string(root["source_sha256"], _SHA256, "semantic_draft_contract_invalid")
    _string(root["revision_id"], _ID, "semantic_draft_contract_invalid")
    _integer(root["draft_revision"], 0, 4_294_967_295, "semantic_draft_contract_invalid")
    provenance = _object(
        root["provenance"],
        frozenset({"source_kind", "application_owned"}),
        "semantic_draft_contract_invalid",
    )
    if provenance != {"source_kind": "studio_semantic_draft", "application_owned": True}:
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    zones = root["zones"]
    flow_paths = root["flow_paths"]
    if not isinstance(zones, list) or len(zones) > MAX_ZONES:
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    if not isinstance(flow_paths, list) or len(flow_paths) > MAX_FLOW_PATHS:
        raise ContamSemanticDraftError("semantic_draft_contract_invalid")
    for zone in zones:
        _zone(zone)
    for flow_path in flow_paths:
        _flow_path(flow_path)
    return root


def canonical_draft_json(value: Mapping[str, Any]) -> bytes:
    return json.dumps(
        value,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
    ).encode("utf-8")


def semantic_draft_sha256(value: Mapping[str, Any]) -> str:
    return hashlib.sha256(canonical_draft_json(value)).hexdigest()


def _geometry_bindings(
    geometry: Mapping[str, Any],
) -> tuple[dict[str, str], dict[str, int], dict[str, tuple[str, int, str | None, str | None, str]]]:
    regions: dict[str, str] = {}
    openings: dict[str, int] = {}
    flow_paths: dict[str, tuple[str, int, str | None, str | None, str]] = {}
    levels = geometry.get("levels")
    if not isinstance(levels, list):
        return regions, openings, flow_paths
    for level in levels:
        if not isinstance(level, dict):
            continue
        level_number = level.get("level_number")
        if isinstance(level_number, bool) or not isinstance(level_number, int):
            continue
        for region in level.get("zone_regions", []):
            if isinstance(region, dict) and isinstance(region.get("id"), str) and isinstance(
                region.get("semantic_zone_id"), str
            ):
                regions[region["id"]] = region["semantic_zone_id"]
        for opening in level.get("openings", []):
            if isinstance(opening, dict) and isinstance(opening.get("id"), str):
                openings[opening["id"]] = level_number
        for anchor in level.get("flow_path_anchors", []):
            if (
                isinstance(anchor, dict)
                and isinstance(anchor.get("semantic_flow_path_id"), str)
                and isinstance(anchor.get("opening_id"), str)
                and (anchor.get("from_zone_id") is None or isinstance(anchor.get("from_zone_id"), str))
                and (anchor.get("to_zone_id") is None or isinstance(anchor.get("to_zone_id"), str))
                and anchor.get("exterior_side") in {"none", "from", "to"}
            ):
                flow_paths[anchor["semantic_flow_path_id"]] = (
                    anchor["opening_id"],
                    level_number,
                    anchor.get("from_zone_id"),
                    anchor.get("to_zone_id"),
                    anchor["exterior_side"],
                )
    return regions, openings, flow_paths


def validate_contam_semantic_draft(value: Any, context: DraftContext) -> DraftValidation:
    try:
        draft = parse_contam_semantic_draft(value)
    except ContamSemanticDraftError as error:
        return DraftValidation("invalid", "", (error.code,))
    diagnostics: set[str] = set()
    if (
        draft["project_session_id"] != context.project_session_id
        or draft["revision_id"] != context.revision_id
        or draft["identity_sha256"].lower() != context.identity_sha256.lower()
        or draft["source_sha256"].lower() != context.source_sha256.lower()
    ):
        diagnostics.add("semantic_draft_context_stale")
    zone_ids: set[str] = set()
    zone_names: set[str] = set()
    region_ids: set[str] = set()
    for zone in draft["zones"]:
        zone_id = zone["id"]
        name = zone["name"].lower()
        region_id = zone["geometry_region_id"]
        if zone_id in zone_ids:
            diagnostics.add("semantic_draft_zone_id_duplicate")
        if name in zone_names:
            diagnostics.add("semantic_draft_zone_name_duplicate")
        if region_id in region_ids:
            diagnostics.add("semantic_draft_zone_region_duplicate")
        zone_ids.add(zone_id)
        zone_names.add(name)
        region_ids.add(region_id)
        if zone["level_number"] not in context.level_numbers:
            diagnostics.add("semantic_draft_level_unknown")
    all_zone_ids = context.existing_zone_ids | zone_ids
    path_ids: set[str] = set()
    opening_ids: set[str] = set()
    for path in draft["flow_paths"]:
        if path["id"] in path_ids:
            diagnostics.add("semantic_draft_flow_path_id_duplicate")
        if path["opening_id"] in opening_ids:
            diagnostics.add("semantic_draft_opening_duplicate")
        path_ids.add(path["id"])
        opening_ids.add(path["opening_id"])
        if path["level_number"] not in context.level_numbers:
            diagnostics.add("semantic_draft_level_unknown")
        if path["flow_element_id"] not in context.supported_flow_element_ids:
            diagnostics.add("semantic_draft_flow_element_unsupported")
        endpoints = (path["from_endpoint"], path["to_endpoint"])
        if all(endpoint["kind"] == "outdoor" for endpoint in endpoints):
            diagnostics.add("semantic_draft_flow_path_outdoor_to_outdoor")
        for endpoint in endpoints:
            if endpoint["kind"] == "zone" and endpoint["zone_id"] not in all_zone_ids:
                diagnostics.add("semantic_draft_endpoint_unknown")
        if (
            endpoints[0]["kind"] == "zone"
            and endpoints[1]["kind"] == "zone"
            and endpoints[0]["zone_id"] == endpoints[1]["zone_id"]
        ):
            diagnostics.add("semantic_draft_flow_path_self_reference")
    if context.geometry is not None:
        geometry_regions, geometry_openings, geometry_flow_paths = _geometry_bindings(context.geometry)
        for zone in draft["zones"]:
            if geometry_regions.get(zone["geometry_region_id"]) != zone["id"]:
                diagnostics.add("semantic_draft_zone_geometry_mismatch")
        for path in draft["flow_paths"]:
            expected_from = path["from_endpoint"]["zone_id"] if path["from_endpoint"]["kind"] == "zone" else None
            expected_to = path["to_endpoint"]["zone_id"] if path["to_endpoint"]["kind"] == "zone" else None
            expected_exterior = (
                "from"
                if path["from_endpoint"]["kind"] == "outdoor"
                else "to" if path["to_endpoint"]["kind"] == "outdoor" else "none"
            )
            if (
                geometry_openings.get(path["opening_id"]) != path["level_number"]
                or geometry_flow_paths.get(path["id"])
                != (
                    path["opening_id"],
                    path["level_number"],
                    expected_from,
                    expected_to,
                    expected_exterior,
                )
            ):
                diagnostics.add("semantic_draft_flow_path_geometry_mismatch")
    return DraftValidation(
        "invalid" if diagnostics else "valid",
        semantic_draft_sha256(draft),
        tuple(sorted(diagnostics)),
    )


def create_empty_contam_semantic_draft(
    *,
    draft_id: str,
    project_session_id: str,
    identity_sha256: str,
    source_sha256: str,
    revision_id: str,
) -> dict[str, Any]:
    draft: dict[str, Any] = {
        "schema_version": SCHEMA_VERSION,
        "status": "available",
        "draft_id": draft_id,
        "project_session_id": project_session_id,
        "identity_sha256": identity_sha256.lower(),
        "source_sha256": source_sha256.lower(),
        "revision_id": revision_id,
        "draft_revision": 0,
        "provenance": {"source_kind": "studio_semantic_draft", "application_owned": True},
        "zones": [],
        "flow_paths": [],
    }
    parse_contam_semantic_draft(draft)
    return draft

File: src/app/geometry/contam-semantic-draft.ts
import { canonicalJson, sha256Text, type BuildingGeometry } from "./geometry-model";

export const CONTAM_SEMANTIC_DRAFT_SCHEMA_VERSION = "contam_semantic_draft.v1" as const;
export const MAX_SEMANTIC_DRAFT_ZONES = 256;
export const MAX_SEMANTIC_DRAFT_FLOW_PATHS = 512;

export interface ContamDraftEndpoint {
  kind: "zone" | "outdoor";
  zone_id: string | null;
}

export interface ContamDraftZone {
  id: string;
  level_number: number;
  name: string;
  display_name: string;
  volume_litres: number;
  volume_basis: "explicit" | "geometry_estimate_confirmed";
  geometry_region_id: string;
  initial_temperature_millikelvin: number;
  initial_pressure_millipascal: number;
}

export interface ContamDraftFlowPath {
  id: string;
  level_number: number;
  opening_id: string;
  from_endpoint: ContamDraftEndpoint;
  to_endpoint: ContamDraftEndpoint;
  flow_element_id: string;
  multiplier_millionths: number;
  x_mm: number;
  y_mm: number;
  relative_height_mm: number;
  direction_degrees: number;
}

export interface ContamSemanticDraft {
  schema_version: typeof CONTAM_SEMANTIC_DRAFT_SCHEMA_VERSION;
  status: "available";
  draft_id: string;
  project_session_id: string;
  identity_sha256: string;
  source_sha256: string;
  revision_id: string;
  draft_revision: number;
  provenance: { source_kind: "studio_semantic_draft"; application_owned: true };
  zones: ContamDraftZone[];
  flow_paths: ContamDraftFlowPath[];
}

export interface ContamSemanticDraftContext {
  projectSessionId: string;
  identitySha256: string;
  sourceSha256: string;
  revisionId: string;
  levelNumbers: ReadonlySet<number>;
  existingZoneIds: ReadonlySet<string>;
  supportedFlowElementIds: ReadonlySet<string>;
  geometry?: BuildingGeometry;
}

export interface ContamSemanticDraftValidation {
  status: "valid" | "invalid";
  draft_sha256: string;
  diagnostics: string[];
}

const ID = /^[A-Za-z0-9_.:-]{1,160}$/;
const SHA256 = /^[A-Fa-f0-9]{64}$/;
const CONTAM_NAME = /^[A-Za-z0-9_.-]{1,15}$/;

function exactKeys(value: unknown, keys: string[]): value is Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) return false;
  const actual = Object.keys(value).sort();
  return actual.length === keys.length && actual.every((key, index) => key === [...keys].sort()[index]);
}

function integerBetween(value: unknown, minimum: number, maximum: number): value is number {
  return Number.isSafeInteger(value) && (value as number) >= minimum && (value as number) <= maximum;
}

function safeDisplayName(value: unknown): value is string {
  return typeof value === "string"
    && value.length >= 1
    && value.length <= 80
    && !/[\u0000-\u001f\u007f]/u.test(value)
    && value.trim() === value;
}

function validEndpoint(value: unknown): value is ContamDraftEndpoint {
  if (!exactKeys(value, ["kind", "zone_id"])) return false;
  return value.kind === "outdoor"
    ? value.zone_id === null
    : value.kind === "zone" && typeof value.zone_id === "string" && ID.test(value.zone_id);
}

function validZone(value: unknown): value is ContamDraftZone {
  if (!exactKeys(value, [
    "display_name", "geometry_region_id", "id", "initial_pressure_millipascal",
    "initial_temperature_millikelvin", "level_number", "name", "volume_basis", "volume_litres",
  ])) return false;
  return typeof value.id === "string" && ID.test(value.id)
    && integerBetween(value.level_number, 1, 10_000)
    && typeof value.name === "string" && CONTAM_NAME.test(value.name)
    && safeDisplayName(value.display_name)
    && integerBetween(value.volume_litres, 1, 1_000_000_000)
    && (value.volume_basis === "explicit" || value.volume_basis === "geometry_estimate_confirmed")
    && typeof value.geometry_region_id === "string" && ID.test(value.geometry_region_id)
    && integerBetween(value.initial_temperature_millikelvin, 173_150, 373_150)
    && integerBetween(value.initial_pressure_millipascal, -1_000_000_000, 1_000_000_000);
}

function validFlowPath(value: unknown): value is ContamDraftFlowPath {
  if (!exactKeys(value, [
    "direction_degrees", "flow_element_id", "from_endpoint", "id", "level_number",
    "multiplier_millionths", "opening_id", "relative_height_mm", "to_endpoint", "x_mm", "y_mm",
  ])) return false;
  return typeof value.id === "string" && ID.test(value.id)
    && integerBetween(value.level_number, 1, 10_000)
    && typeof value.opening_id === "string" && ID.test(value.opening_id)
    && validEndpoint(value.from_endpoint)
    && validEndpoint(value.to_endpoint)
    && typeof value.flow_element_id === "string" && ID.test(value.flow_element_id)
    && integerBetween(value.multiplier_millionths, 1, 1_000_000_000_000)
    && integerBetween(value.x_mm, -1_000_000_000, 1_000_000_000)
    && integerBetween(value.y_mm, -1_000_000_000, 1_000_000_000)
    && integerBetween(value.relative_height_mm, 0, 100_000_000)
    && integerBetween(value.direction_degrees, -1, 359);
}

export function isContamSemanticDraft(value: unknown): value is ContamSemanticDraft {
  if (!exactKeys(value, [
    "draft_id", "draft_revision", "flow_paths", "identity_sha256", "project_session_id",
    "provenance", "revision_id", "schema_version", "source_sha256", "status", "zones",
  ])) return false;
  return value.schema_version === CONTAM_SEMANTIC_DRAFT_SCHEMA_VERSION
    && value.status === "available"
    && typeof value.draft_id === "string" && ID.test(value.draft_id)
    && typeof value.project_session_id === "string" && ID.test(value.project_session_id)
    && typeof value.identity_sha256 === "string" && SHA256.test(value.identity_sha256)
    && typeof value.source_sha256 === "string" && SHA256.test(value.source_sha256)
    && typeof value.revision_id === "string" && ID.test(value.revision_id)
    && integerBetween(value.draft_revision, 0, 4_294_967_295)
    && exactKeys(value.provenance, ["application_owned", "source_kind"])
    && value.provenance.source_kind === "studio_semantic_draft"
    && value.provenance.application_owned === true
    && Array.isArray(value.zones) && value.zones.length <= MAX_SEMANTIC_DRAFT_ZONES && value.zones.every(validZone)
    && Array.isArray(value.flow_paths) && value.flow_paths.length <= MAX_SEMANTIC_DRAFT_FLOW_PATHS && value.flow_paths.every(validFlowPath);
}

export function semanticDraftSha256(draft: ContamSemanticDraft): string {
  return sha256Text(canonicalJson(draft));
}

export function validateContamSemanticDraft(
  value: unknown,
  context: ContamSemanticDraftContext,
): ContamSemanticDraftValidation {
  if (!isContamSemanticDraft(value)) {
    return { status: "invalid", draft_sha256: "", diagnostics: ["semantic_draft_contract_invalid"] };
  }
  const draft = value;
  const diagnostics = new Set<string>();
  if (draft.project_session_id !== context.projectSessionId
    || draft.revision_id !== context.revisionId
    || draft.identity_sha256.toLowerCase() !== context.identitySha256.toLowerCase()
    || draft.source_sha256.toLowerCase() !== context.sourceSha256.toLowerCase()) {
    diagnostics.add("semantic_draft_context_stale");
  }
  const zoneIds = new Set<string>();
  const zoneNames = new Set<string>();
  const regionIds = new Set<string>();
  for (const zone of draft.zones) {
    if (zoneIds.has(zone.id)) diagnostics.add("semantic_draft_zone_id_duplicate");
    if (zoneNames.has(zone.name.toLowerCase())) diagnostics.add("semantic_draft_zone_name_duplicate");
    if (regionIds.has(zone.geometry_region_id)) diagnostics.add("semantic_draft_zone_region_duplicate");
    zoneIds.add(zone.id);
    zoneNames.add(zone.name.toLowerCase());
    regionIds.add(zone.geometry_region_id);
    if (!context.levelNumbers.has(zone.level_number)) diagnostics.add("semantic_draft_level_unknown");
  }
  const allZoneIds = new Set([...context.existingZoneIds, ...zoneIds]);
  const pathIds = new Set<string>();
  const openingIds = new Set<string>();
  for (const path of draft.flow_paths) {
    if (pathIds.has(path.id)) diagnostics.add("semantic_draft_flow_path_id_duplicate");
    if (openingIds.has(path.opening_id)) diagnostics.add("semantic_draft_opening_duplicate");
    pathIds.add(path.id);
    openingIds.add(path.opening_id);
    if (!context.levelNumbers.has(path.level_number)) diagnostics.add("semantic_draft_level_unknown");
    if (!context.supportedFlowElementIds.has(path.flow_element_id)) diagnostics.add("semantic_draft_flow_element_unsupported");
    const endpoints = [path.from_endpoint, path.to_endpoint];
    if (endpoints.every((endpoint) => endpoint.kind === "outdoor")) diagnostics.add("semantic_draft_flow_path_outdoor_to_outdoor");
    for (const endpoint of endpoints) {
      if (endpoint.kind === "zone" && (!endpoint.zone_id || !allZoneIds.has(endpoint.zone_id))) {
        diagnostics.add("semantic_draft_endpoint_unknown");
      }
    }
    if (path.from_endpoint.kind === "zone" && path.to_endpoint.kind === "zone"
      && path.from_endpoint.zone_id === path.to_endpoint.zone_id) {
      diagnostics.add("semantic_draft_flow_path_self_reference");
    }
  }
  if (context.geometry) {
    const geometryRegions = new Map(context.geometry.levels.flatMap((level) => level.zone_regions.map((region) => [region.id, region] as const)));
    const geometryOpenings = new Map(context.geometry.levels.flatMap((level) => level.openings.map((opening) => [opening.id, { opening, levelNumber: level.level_number }] as const)));
    const geometryFlowPaths = new Map(context.geometry.levels.flatMap((level) => level.flow_path_anchors.map((anchor) => [anchor.semantic_flow_path_id, { anchor, levelNumber: level.level_number }] as const)));
    for (const zone of draft.zones) {
      const region = geometryRegions.get(zone.geometry_region_id);
      if (!region || region.semantic_zone_id !== zone.id) diagnostics.add("semantic_draft_zone_geometry_mismatch");
    }
    for (const path of draft.flow_paths) {
      const opening = geometryOpenings.get(path.opening_id);
      const binding = geometryFlowPaths.get(path.id);
      const expectedFrom = path.from_endpoint.kind === "zone" ? path.from_endpoint.zone_id : null;
      const expectedTo = path.to_endpoint.kind === "zone" ? path.to_endpoint.zone_id : null;
      const expectedExterior = path.from_endpoint.kind === "outdoor"
        ? "from"
        : path.to_endpoint.kind === "outdoor" ? "to" : "none";
      if (!opening
        || opening.levelNumber !== path.level_number
        || !binding
        || binding.levelNumber !== path.level_number
        || binding.anchor.opening_id !== path.opening_id
        || binding.anchor.from_zone_id !== expectedFrom
        || binding.anchor.to_zone_id !== expectedTo
        || binding.anchor.exterior_side !== expectedExterior) {
        diagnostics.add("semantic_draft_flow_path_geometry_mismatch");
      }
    }
  }
  return {
    status: diagnostics.size ? "invalid" : "valid",
    draft_sha256: semanticDraftSha256(draft),
    diagnostics: [...diagnostics].sort(),
  };
}

export function createEmptyContamSemanticDraft(context: Omit<ContamSemanticDraftContext, "levelNumbers" | "existingZoneIds" | "supportedFlowElementIds" | "geometry">, draftId: string): ContamSemanticDraft {
  return {
    schema_version: CONTAM_SEMANTIC_DRAFT_SCHEMA_VERSION,
    status: "available",
    draft_id: draftId,
    project_session_id: context.projectSessionId,
    identity_sha256: context.identitySha256.toLowerCase(),
    source_sha256: context.sourceSha256.toLowerCase(),
    revision_id: context.revisionId,
    draft_revision: 0,
    provenance: { source_kind: "studio_semantic_draft", application_owned: true },
    zones: [],
    flow_paths: [],
  };
}
